src/controllers/apps/chat-app/message.controllers.js

```diff
@@ -154,5 +158,128 @@ const sendMessage = asyncHandler(async (req, res) => {
 
-export { getAllMessages, sendMessage };
+  const { chatId, messageId, attachmentId } = req.params;
+
+  //Find the chat based on chatId
+
+  const chat = await Chat.findOne({
+    _id: new mongoose.Types.ObjectId(chatId),
+  });
+
+  if (!chat) {
+    throw new ApiError(404, "Chat does not exist");
+  }
+  //Find the message based on message id
+
+  const message = await ChatMessage.findOne({
+    _id: new mongoose.Types.ObjectId(messageId),
+  });
+
+  if (!message) {
+    throw new ApiError(404, "Message does not exist");
+  }
+
+  // Check if user is the sender of the message
+  if (message.sender.toString() !== req.user._id.toString()) {
+    throw new ApiError(
+      401,
+      "You are not authorised to delete the message ,you are not the sender"
+    );
+  }
+
+  //Checking if 15 mins has passed since the message sent or not
+
+  const currentTime = new Date();
+  const messageCreatedAt = message.createdAt;
+  const timeDifferenceMinutes = (currentTime - messageCreatedAt) / (1000 * 60); //calculating the time difference in minutes
+
+  if (timeDifferenceMinutes > 15) {
+    throw new ApiError(
+      400,
+      "15 mins have passed you cannot delete this message"
+    );
+  }
+
+  // If frontend sends the attachment id and message has more than one attachment
+  //Then delete only the attachment sent by the user otherwise delete the whole message
+  if (attachmentId && message.attachments.length > 1) {
+    const attachmentIndex = message.attachments.findIndex(
+      (attachment) => attachment._id.toString() === attachmentId
+    );
+
+    if (attachmentIndex === -1) {
+      throw new ApiError(404, "Attachment not found");
+    }
+
+    const deletedAttachment = message.attachments.splice(attachmentIndex, 1);
+    removeLocalFile(deletedAttachment[0].localPath);
+
+    await message.save({ validateBeforeSave: false });
+    // logic to emit socket event about the message attachment deleted  to the other participants
+    chat.participants.forEach((participantObjectId) => {
+      // here the chat is the raw instance of the chat in which participants is the array of object ids of users
+      // avoid emitting event to the user who is deleting the message
+      if (participantObjectId.toString() === req.user._id.toString()) return;
+
+      // emit the delete message attachment event to the other participants frontend with messageId as the payload
+      emitSocketEvent(
+        req,
+        participantObjectId.toString(),
+        ChatEventEnum.MESSAGE_DELETED_EVENT,
+        message._id
+      );
+    });
+
+    return res
+      .status(200)
+      .json(new ApiResponse(200, {}, "Message attachment deleted succesfully"));
+  }
+  //If the message is attachment  remove the attachments from the server
+  if (message.attachments.length > 0) {
+    message.attachments.map((asset) => {
+      removeLocalFile(asset.localPath);
+    });
+  }
+  //deleting the message from DB
+  try {
```

Comment:
Why are we using `try catch` here? we are already wrapping whole controller in `asyncHandler` no need to add try catch.